### ISSUE TITLE
Is it possible to support GHC 7.8?

### ISSUE THREAD

**Issue by asr** [int-e/haskell-src-exts-simple#1]
I would like to use `haskell-src-exts-simple` in [Agda](https://github.com/agda/agda/), which currently supports GHC >= 7.6. We will remove the support for GHC 7.6 this year (maybe).

From the [README](https://github.com/int-e/haskell-src-exts-simple/blob/2353fefc16508458225f8e45c4da3a46eb633603/COMPATIBILITY.md#ghc-versions), I know that `haskell-src-exts-simple` supports GHC >= 7.10. Is it possible to extend the support for GHC 7.8?

**Comment by int-e** [comment]
The only reason for the ghc 7.10 requirement right now is support for literals (`Language.Haskell..Exts.Simple.Literal`), which currently use explicitly bidirectional pattern synonyms:

```
type Literal = H.Literal ()

pattern Char a <- H.Char () (a :: Char) _ :: Literal
    where Char a = H.Char () a [a]
[...]
```

I can imagine replacing this by

```
type Literal = H.Literal ()

pattern Char a <- H.Char () (a :: Char) _ :: Literal
charL :: Char -> Literal
charL a = H.Char () a [a]
```

etc. which would mean that all the `Literal` constructors could only be used for matching.

**Comment by int-e** [comment]
I think I'll make this change (this will benefit lambdabot as well).

An open question is whether to use CPP magic to provide bidirectional patterns for ghc 7.10 and beyond. There is the usual tradeoff: On the one hand the interface would be nicer for newer ghcs; on the other hand, people developing with ghc-7.10 may be surprised to find their code not working when compiled with ghc-7.8.

**Comment by int-e** [comment]
There is another option: Export the `Literal` constructors _with_ their extraneous `String` argument. Do you have any preferences?

Note that these values are ignored by the (non-exact) pretty-printer, so they really are extraneous in the context of haskell-src-exts-simple: [Language.Haskell.Exts.Pretty#L914-L925](https://github.com/haskell-suite/haskell-src-exts/blob/9ee3b655b2de62ed5584457054704985a3ef35cb/src/Language/Haskell/Exts/Pretty.hs#L914-L925)

**Comment by int-e** [comment]
Unless there are any objections I'll go with the CPP hack version (see https://github.com/int-e/haskell-src-exts-simple/commit/5ab83bdf119cca6361799bbf11ea65ca74848f91)

**Comment by int-e** [comment]
Released as version 1.18.0.1.
